File: backend/internal/repository/hug/streak.go
package hug

import (
	"context"
	"time"

	"go-service-template/internal/db/sqlc/storage"
	"go-service-template/internal/models"
	"go-service-template/internal/repository"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgtype"
)

func (r *repo) GetPairStreak(ctx context.Context, userA, userB uuid.UUID) (*models.PairStreak, error) {
	q := repository.Queries(ctx, r.q)

	row, err := q.GetPairStreak(ctx, storage.GetPairStreakParams{
		UserA: userA,
		UserB: userB,
	})
	if err != nil {
		if err == pgx.ErrNoRows {
			return nil, nil
		}
		return nil, err
	}

	return toModelPairStreak(row.UserAID, row.UserBID, row.CurrentStreak, row.BestStreak,
		row.LastStreakDate, row.AHuggedToday, row.BHuggedToday, row.TodayDate), nil
}

func (r *repo) UpsertPairStreak(ctx context.Context, streak *models.PairStreak) (*models.PairStreak, error) {
	q := repository.Queries(ctx, r.q)

	params := storage.UpsertPairStreakParams{
		UserA:         streak.UserAID,
		UserB:         streak.UserBID,
		CurrentStreak: streak.CurrentStreak,
		BestStreak:    streak.BestStreak,
		AHuggedToday:  streak.AHuggedToday,
		BHuggedToday:  streak.BHuggedToday,
		TodayDate:     pgtype.Date{Time: streak.TodayDate, Valid: true},
	}
	if streak.LastStreakDate != nil {
		params.LastStreakDate = pgtype.Date{Time: *streak.LastStreakDate, Valid: true}
	}

	row, err := q.UpsertPairStreak(ctx, params)
	if err != nil {
		return nil, err
	}

	return toModelPairStreak(row.UserAID, row.UserBID, row.CurrentStreak, row.BestStreak,
		row.LastStreakDate, row.AHuggedToday, row.BHuggedToday, row.TodayDate), nil
}

func (r *repo) GetUserTopStreaks(ctx context.Context, userID uuid.UUID, limit int32) ([]*models.TopStreakEntry, error) {
	q := repository.Queries(ctx, r.q)

	rows, err := q.GetUserTopStreaks(ctx, storage.GetUserTopStreaksParams{
		UserID: userID,
		Lim:    limit,
	})
	if err != nil {
		return nil, err
	}

	today := time.Now().UTC().Truncate(24 * time.Hour)

	result := make([]*models.TopStreakEntry, 0, len(rows))
	for _, row := range rows {
		streak := evaluateStreakFreshness(row.CurrentStreak, row.BestStreak,
			row.LastStreakDate, row.AHuggedToday, row.BHuggedToday, row.TodayDate, today)

		// Skip entries that are no longer active after freshness evaluation
		if streak.CurrentStreak <= 0 {
			continue
		}

		tier := models.ComputeStreakTier(streak.CurrentStreak)

		var otherUserID uuid.UUID
		if row.UserAID == userID {
			otherUserID = row.UserBID
		} else {
			otherUserID = row.UserAID
		}

		var displayName *string
		if row.OtherDisplayName.Valid {
			displayName = &row.OtherDisplayName.String
		}
		var gender *string
		if row.OtherGender.Valid {
			gender = &row.OtherGender.String
		}

		result = append(result, &models.TopStreakEntry{
			UserID:        otherUserID,
			Username:      row.OtherUsername,
			DisplayName:   displayName,
			Gender:        gender,
			CurrentStreak: streak.CurrentStreak,
			BestStreak:    streak.BestStreak,
			TierLevel:     tier.Level,
			TierName:      tier.Name,
			TierKey:       tier.Key,
		})
	}

	return result, nil
}

func (r *repo) GetPairStreakCalendar(ctx context.Context, userA, userB uuid.UUID, since time.Time) ([]*models.StreakCalendarDay, error) {
	q := repository.Queries(ctx, r.q)

	rows, err := q.GetPairStreakCalendar(ctx, storage.GetPairStreakCalendarParams{
		UserA: userA,
		UserB: userB,
		Since: pgtype.Timestamptz{Time: since, Valid: true},
	})
	if err != nil {
		return nil, err
	}

	result := make([]*models.StreakCalendarDay, len(rows))
	for i, row := range rows {
		result[i] = &models.StreakCalendarDay{
			Date:      row.HugDate.Time,
			HugCount:  row.HugCount,
			Completed: row.Completed,
		}
	}
	return result, nil
}

// toModelPairStreak converts DB types to domain PairStreak.
func toModelPairStreak(userAID, userBID uuid.UUID, currentStreak, bestStreak int32,
	lastStreakDate pgtype.Date, aHuggedToday, bHuggedToday bool, todayDate pgtype.Date) *models.PairStreak {
	ps := &models.PairStreak{
		UserAID:       userAID,
		UserBID:       userBID,
		CurrentStreak: currentStreak,
		BestStreak:    bestStreak,
		AHuggedToday:  aHuggedToday,
		BHuggedToday:  bHuggedToday,
		TodayDate:     todayDate.Time,
	}
	if lastStreakDate.Valid {
		t := lastStreakDate.Time
		ps.LastStreakDate = &t
	}
	return ps
}

// evaluateStreakFreshness checks if the streak data is stale and computes the effective current values.
// This implements lazy evaluation: if today_date in the DB is not today, we figure out what really happened.
func evaluateStreakFreshness(currentStreak, bestStreak int32, lastStreakDate pgtype.Date,
	aHuggedToday, bHuggedToday bool, todayDate pgtype.Date, today time.Time) struct {
	CurrentStreak int32
	BestStreak    int32
} {
	dbToday := todayDate.Time
	// If the record's today_date is actually today, return as-is
	if dbToday.Equal(today) {
		return struct {
			CurrentStreak int32
			BestStreak    int32
		}{currentStreak, bestStreak}
	}

	// The record is stale. Check if the previous day (todayDate) was completed.
	yesterday := today.AddDate(0, 0, -1)
	if aHuggedToday && bHuggedToday && dbToday.Equal(yesterday) {
		// Previous day was completed and it was yesterday — streak continues
		return struct {
			CurrentStreak int32
			BestStreak    int32
		}{currentStreak, bestStreak}
	}

	// If last_streak_date was yesterday, the streak is still valid (just no hugs today yet)
	if lastStreakDate.Valid && lastStreakDate.Time.Equal(yesterday) {
		return struct {
			CurrentStreak int32
			BestStreak    int32
		}{currentStreak, bestStreak}
	}

	// If last_streak_date was today, streak is valid
	if lastStreakDate.Valid && lastStreakDate.Time.Equal(today) {
		return struct {
			CurrentStreak int32
			BestStreak    int32
		}{currentStreak, bestStreak}
	}

	// Otherwise streak is broken
	return struct {
		CurrentStreak int32
		BestStreak    int32
	}{0, bestStreak}
}

File: backend/src/models/streak.rs
use chrono::NaiveDate;
use uuid::Uuid;

#[derive(Debug, Clone)]
pub struct PairStreak {
    pub user_a_id: Uuid,
    pub user_b_id: Uuid,
    pub current_streak: i32,
    pub best_streak: i32,
    pub last_streak_date: Option<NaiveDate>,
    pub a_hugged_today: bool,
    pub b_hugged_today: bool,
    pub today_date: NaiveDate,
}

#[derive(Debug, Clone, Copy)]
pub struct StreakTier {
    pub level: i32,
    pub name: &'static str,
    pub key: &'static str,
    pub min_days: i32,
}

pub const STREAK_TIERS: &[StreakTier] = &[
    StreakTier {
        level: 6,
        name: "Легендарная",
        key: "legendary",
        min_days: 90,
    },
    StreakTier {
        level: 5,
        name: "Обсидиановая",
        key: "obsidian",
        min_days: 60,
    },
    StreakTier {
        level: 4,
        name: "Алмазная",
        key: "diamond",
        min_days: 30,
    },
    StreakTier {
        level: 3,
        name: "Сапфировая",
        key: "sapphire",
        min_days: 21,
    },
    StreakTier {
        level: 2,
        name: "Рубиновая",
        key: "ruby",
        min_days: 14,
    },
    StreakTier {
        level: 1,
        name: "Изумрудная",
        key: "emerald",
        min_days: 7,
    },
    StreakTier {
        level: 0,
        name: "",
        key: "",
        min_days: 0,
    },
];

pub fn compute_streak_tier(streak_days: i32) -> StreakTier {
    for tier in STREAK_TIERS {
        if streak_days >= tier.min_days {
            return *tier;
        }
    }
    STREAK_TIERS[STREAK_TIERS.len() - 1]
}

#[derive(Debug, Clone)]
pub struct StreakInfo {
    pub current_streak: i32,
    pub best_streak: i32,
    pub tier_level: i32,
    pub tier_name: String,
    pub tier_key: String,
    pub next_tier_at: Option<i32>,
    pub a_hugged_today: bool,
    pub b_hugged_today: bool,
}

pub fn compute_streak_info(streak: Option<&PairStreak>) -> StreakInfo {
    let Some(streak) = streak else {
        return StreakInfo {
            current_streak: 0,
            best_streak: 0,
            tier_level: 0,
            tier_name: String::new(),
            tier_key: String::new(),
            next_tier_at: None,
            a_hugged_today: false,
            b_hugged_today: false,
        };
    };
    let tier = compute_streak_tier(streak.current_streak);
    let mut info = StreakInfo {
        current_streak: streak.current_streak,
        best_streak: streak.best_streak,
        tier_level: tier.level,
        tier_name: tier.name.to_string(),
        tier_key: tier.key.to_string(),
        next_tier_at: None,
        a_hugged_today: streak.a_hugged_today,
        b_hugged_today: streak.b_hugged_today,
    };
    for t in STREAK_TIERS.iter().rev() {
        if t.min_days > streak.current_streak {
            info.next_tier_at = Some(t.min_days);
            break;
        }
    }
    info
}

#[derive(Debug, Clone)]
pub struct TopStreakEntry {
    pub user_id: Uuid,
    pub username: String,
    pub display_name: Option<String>,
    pub gender: Option<String>,
    pub current_streak: i32,
    pub best_streak: i32,
    pub tier_level: i32,
    pub tier_name: String,
    pub tier_key: String,
}

#[derive(Debug, Clone)]
pub struct StreakCalendarDay {
    pub date: NaiveDate,
    pub hug_count: i64,
    pub completed: bool,
}
